### PR TITLE
Add share link entry

### PR DESCRIPTION
**Pull request by skjnldsv**
Closes https://github.com/nextcloud/server/issues/28376

### CODE REVIEW

**Review by eneiluj — COMMENTED** [review]
Code looks good. Too bad we can't try it because `ExternalShareAction is not allowed inside the Actions component`.

One question though: Is the `DownloadLimitAction` displayed even if the limit is disabled?

File: .eslintrc.js [hunk 1]
```diff
@@ -1,3 +1,6 @@
 module.exports = {
+	globals: {
+		appName: true,
```
Comment (Pytal) [inline .eslintrc.js:3]: After checking the docs https://eslint.org/docs/user-guide/configuring/language-options#using-configuration-files-1 shouldn't this be set to `'readonly'`?

**Comment by PVince81** [comment]
- [ ] add README with notes about how to set the counter as I couldn't find it

**Comment by PVince81** [comment]
or it's a bug... I'm currently on server master 995aa6518357b9f371ced5ba8941dfd2a02b0d97 and this PR.

I can't see any option to set the download limit:

image

also I didn't find any option in the admin's share settings, in case the admin needs to enable it.

so either it's not working or it's well hidden ?

**Comment by PVince81** [comment]
- [ ] REQUIRES  https://github.com/nextcloud/server/pull/28835 or https://github.com/nextcloud/server/pull/28842

File: lib/Listener/ShareLinkAccessedListener.php [hunk 2]
```diff
@@ -65,18 +66,27 @@ public function handle(Event $event): void {
...
-		$limit = $this->config->getAppValue(Application::APP_ID, 'download_limit_' . $token, -1);
+		try {
+			$shareLimit = $this->mapper->get($token);
+			$limit = $shareLimit->getLimit();
 
-		if ($limit > 0) {
-			$count = $this->config->getAppValue(Application::APP_ID, 'download_count_' . $token, 0) + 1;
-			$this->config->setAppValue(Application::APP_ID, 'download_count_' . $token, $count);
+			// Increment this download event
+			$downloads = $shareLimit->getDownloads() + 1;
```
Comment (PVince81) [inline lib/Listener/ShareLinkAccessedListener.php:74]: potential concurrency concern, I've raised it https://github.com/nextcloud/files_downloadlimit/issues/8 as it doesn't look that critical and rather unlikely to be triggered by a regular user

**Review by PVince81 — APPROVED** [review]
:+1: code looks fine to me so far

I was not able to test due to dependencies.
Also see comment + ticket about potential concurrency concern.

**Comment by PVince81** [comment]
I guess we shouldn't merge until this actually works after resolving the related PRs

**Comment by szaimen** [comment]
> I can't see any option to set the download limit:
> 
> image
> 
> also I didn't find any option in the admin's share settings, in case the admin needs to enable it.
> 
> so either it's not working or it's well hidden ?

Same for me. 

<details>
<summary>Tested with</summary>

```
docker run -d \
-e SERVER_BRANCH=feat/ExternalShareActions-API \
-e DOWNLOADLIMIT_BRANCH=feat/front-menu-entry \
-p 8443:443 \
-e TRUSTED_DOMAIN=192.168.178.34 \
--name nextcloud-easy-test \
--user=www-data \
ghcr.io/szaimen/nextcloud-easy-test:latest
```

</details>

**Comment by szaimen** [comment]
> Just use server's `feat/ExternalShareActions-API-bis` branch and you'll be able to see the sharing entries.

Can confirm that it shows up with your branch but that has its own problems...  So probably we need to decide first which attempt to use and then test it and fix the issues there...

File: src/public.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+console.debug('[DEBUG]', appName, { limit, downloads })
+
+window.addEventListener('DOMContentLoaded', function() {
+	if (limit > 0) {
+		const count = limit - downloads
+		const container = document.getElementById('header-primary-action')
+		const span = document.createElement('span')
+
+		span.style = 'color: var(--color-primary-text); padding: 0 10px;'
+		span.innerText = n('files_downloadlimit', '1 more download allowed', '{count} more downloads allowed', count, { count })
```
Comment (Pytal) [inline src/public.js:35]: Fine with me 👍 would suggest the same as in https://github.com/nextcloud/files_downloadlimit/pull/7#discussion_r709474050 to have `1 remaining download allowed` for singular and `{count} remaining downloads allowed` for the plural version :)

**Review by Pytal — APPROVED** [review]

**Comment by skjnldsv** [comment]
https://github.com/nextcloud/nextcloud-vue/pull/2252

**Pull request merged**